cpp/tests/utilities/column_utilities.cu

```diff
@@ -57,8 +58,65 @@ public:
   }
 };
 
-void expect_columns_equal(cudf::column_view const& lhs, cudf::column_view const& rhs,
-                          bool print_all_differences) {
+class corresponding_rows_not_equivalent {
+  table_device_view d_lhs; 
+  table_device_view d_rhs;
+public:
+  corresponding_rows_not_equivalent(table_device_view d_lhs, table_device_view d_rhs)
+  : d_lhs(d_lhs),
+    d_rhs(d_rhs),
+    comp(d_lhs, d_rhs)
+  {
+    CUDF_EXPECTS(d_lhs.num_columns() == 1 and d_rhs.num_columns() == 1,
+                 "Unsupported number of columns");
+  }
+  
+  struct typed_element_not_equivalent {
+    template <typename T>
+    __device__ std::enable_if_t<std::is_floating_point<T>::value, bool>
+    operator()(column_device_view const& lhs,
+               column_device_view const& rhs,
+               size_type index)
+    {
+      if (lhs.is_valid(index) and rhs.is_valid(index)) {
+        int ulp = 4; // value taken from google test
+        T x = lhs.element<T>(index);
+        T y = rhs.element<T>(index);
+        return std::abs(x-y) > std::numeric_limits<T>::epsilon() * std::abs(x+y) * ulp
```

Comment:
Where does this logic come from? It looks like it's mixing an epsilon _and_ units of least precision, which doesn't quite make sense to me. 

See https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/ for a discussion on using a relative episilon _or_ a fixed number of units of least precision. I'd opt for the latter.

Reply:
Got it from here: https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

Reply:
Hm, still looks kinda funny to me, but it's been too long since I've taken a numerical analysis class to dive any further into it. We'll trust that since it's on cppreference it should be "good enough".